### PR TITLE
Try new syntax for dynamic selectors

### PR DESCRIPTION
**Pull request by oandregal**
Related:

- [theme.json spec](https://developer.wordpress.org/block-editor/developers/themes/theme-json/)
- try encoding dynamic selectors as block variations https://github.com/WordPress/gutenberg/pull/22805
- https://github.com/WordPress/gutenberg/pull/22518#discussion_r429180157
- https://github.com/WordPress/gutenberg/pull/22698#issuecomment-635352451

Blocks that support the Global Styles system need a way to be converted to a CSS selector. At the moment, the way we do it is by using this algorithm:

1. If the block doesn't define a `__experimentalSelector` within its block.json, the selector is built based on its name: `.wp-block-${ blockName }`.Example: `core/group` maps to `.wp-block-group`.

2. If the block does define a `__experimentalSelector` key within its block.json and it is a string, the selector is the value of `__experimentalSelector`. Example: `core/paragraph` maps to `p` because it declares support by `"__experimentalSelector": "p"`.

3. If the block does define a `__experimentalSelector` key within its block.json and it is an object, the block has as many selectors as object entries. Example: `core/heading` can be `core/heading/h1`, `core/heading/h2`, ..., `core/heading/h6` and each one has its own selector that is defined this way:

```
"__experimentalSelector": {
    "core/heading/h1": "h1",
    "core/heading/h2": "h2",
    "core/heading/h3": "h3",
    "core/heading/h4": "h4",
    "core/heading/h5": "h5",
    "core/heading/h6": "h6"
}
```

This works fine for the static nature of server-side global styles (it doesn't use the block values, only its description). However, for the dynamic nature of the client-side (particularly [feature detection](https://github.com/WordPress/gutenberg/pull/24416)) we need a way to map the specific instance of a block to a particular selector. Example: the core/heading block when its level is 3 should be detected as `core/heading/h3`, while it should be detected as `core/heading/h2` when its level is 2.

This PR proposes a new syntax for dynamic selectors:

```
"__experimentalSelector": {
    "attribute": "level",
    "expression": "h${attribute}",
    "range": ["h1", "h2", "h3", "h4", "h5", "h6"]
}
```

The `attribute` and `expression` would be used client-side (or in mobile) to construct the selector based on dynamic data (the value of the `level` attribute). The `range` would be used server-side to process the block.

Note that I haven't implemented the change yet because I want to gather thoughts first.

### CODE REVIEW

**Comment by jorgefilipecosta** [comment]
Hi @nosolosw,
What if relate these CSS selectors to the block variations concept:

1. We would only have simple selectors but block variations could have a CSS selector.
2. For each heading level h1,h2,h3,etc... we would have a block variation, I guess we could add a mechanism to hide variations from the inserter to not pollute it.
3. To identify if a block instance matches a selector if we would identify if the block instance matches a block variation.
4. To know if a block instance matches a variation we could check if all the attributes defined in the variation have the same value in the current block instance.

**Comment by nosolosw** [comment]
Hey Jorge, for context, I'm sharing a little exploration I did about using block variations for this https://github.com/WordPress/gutenberg/pull/22805

If I understood correctly [this proposal](https://github.com/WordPress/gutenberg/pull/24423#issuecomment-680052590), this is what would happen:

1. The user adds the core/heading block as an h2 block variation. At this point, the attribute `level` is 2, and the selector is `h2`.
2. The user changes the block level to be 3. At this point, the attribute `level` is 3, but the selector is still `h2`.

Is that correct? Let me share my raw thinking on this:

- Do we serialize the selector value as `h2`?
- How do we know that we should compute a selector (ex: in the case of heading) or use its value directly (ex: in the case of paragraph) if both are string selectors? I guess we check if the block has variations? If so, can a block define a variation but still use a selector directly without computing it?
- What if a block defines variations with attributes other than the essential to the selector computation? Ex: let's say the core heading defines 6 variations (h1, ..., h6) and each one of them comes with a different value for `level` but also with a different value for `style.fontSize`. How do we know that we only need to take the `level` attribute into account to compute the selector?

I guess I'm struggling to see the value of using block variations _in addition to_ the runtime value of attributes to compute the selector value. :thinking: Computing the selector exclusively from the runtime value of attributes seems simpler (conceptually and from an implementation point of view).

File: packages/block-library/src/post-title/block.json [hunk 1]
```diff
@@ -23,13 +23,9 @@
 		"__experimentalFontSize": true,
 		"__experimentalLineHeight": true,
 		"__experimentalSelector": {
-			"core/post-title/h1": "h1",
```
Comment (jorgefilipecosta) [inline packages/block-library/src/post-title/block.json:26]: It seems we have an issue both post-title and heading block map to the same CSS selectors so CSS output from one block will overwrite the CSS output of the other blocks. One post title the selector should probably also include a class.

**Comment by jorgefilipecosta** [comment]
Hi @nosolosw,

Regarding the possible alternative using the block variations API,

> The user changes the block level to be 3. At this point, the attribute level is 3, but the selector is still h2.

No, in that case, the selector would be h3. The way to match if a variation is being used or not is not by what variation was inserted. We could use a rule for example a given variation is active if all the selectors the variation defines contain the same value in the current block instance.
We would not serialize any selector or have any static information stored.


> I guess we check if the block has variations? If so, can a block define a variation but still use a selector directly without computing it?

We would only use a variation selector if the variation provides a selector, otherwise, we would normally use the block selector.


> What if a block defines variations with attributes other than the essential to the selector computation? Ex: let's say the core heading defines 6 variations (h1, ..., h6) and each one of them comes with a different value for level but also with a different value for style.fontSize. How do we know that we only need to take the level attribute into account to compute the selector?


This case would be the most problematic with the variations approach. If that's the case I guess,  one would need to specify some variations that are hidden, that just contain attributes relevant to compute the selector. But I guess this case would not be very common.




> I guess I'm struggling to see the value of using block variations in addition to the runtime value of attributes to compute the selector value. 🤔 Computing the selector exclusively from the runtime value of attributes seems simpler (conceptually and from an implementation point of view).

The value of using variations is that block authors would not need to learn a new API with a custom syntax they would simply use the variations API that already exists and use the normal __experimentalSelector option that already exists for blocks.

**Comment by jorgefilipecosta** [comment]
Regarding the API being proposed on this PR,

```
"__experimentalSelector": {
    "attribute": "level",
    "expression": "h${attribute}",
    "range": ["h1", "h2", "h3", "h4", "h5", "h6"]
}
```

What would happen if the way to compute the selector needed more than one attribute?
I guess our syntax could use the attribute name inline e.g: "h${level}", so we can reference multiple attributes.


What if we have a block that uses an isInline boolean attribute when true needs a "span.wp-block-x" selector and when false needs a "div.wp-block-x" selector. Or an isVideo that is used when we need a video element but otherwise we use an image selector?

Ideally, we would have some function that given the attributes returns the selector, but that would not allow us to know the selector on the server which may be needed in the future. So I think we should follow a path that allows us to know the selector on the server and the client.

**Comment by nosolosw** [comment]
Going to close this in favor of https://github.com/WordPress/gutenberg/pull/26945 which is simpler!

**Pull request closed without merging**